File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Effect/EffectKit.ts
import { TimeKit } from "../../../../Core/Utils/TimeKit";
import { AbstractEffect } from "./AbstractEffect";


export default class EffectKit{
    private static inst: EffectKit;
    public static Inst(): EffectKit {
        if (!EffectKit.inst) {
            EffectKit.inst = new EffectKit();
        }
        return EffectKit.inst;
    }
    private waiting : boolean;
    private AsyncEffect : Array<AbstractEffect> = new Array<AbstractEffect>();
    private SyncEffect : Array<AbstractEffect> = new Array<AbstractEffect>();
    public Play(effect: AbstractEffect, sync:boolean = true)
    {
        if(sync)
            this.SyncEffect.push(effect);
        else
            this.AsyncEffect.push(effect);
    }

    public Update(){
        for (let i = this.AsyncEffect.length - 1; i >= 0; i--) {
            let node = this.AsyncEffect[i];
            try{
                let result = node.Update();
                if(!node.IsDone)
                {
                    node.Duration+=TimeKit.DeltaTime;
                    break;
                }
                else
                {
                    this.AsyncEffect.splice(i,1);
                }
            }
            catch{
                this.AsyncEffect.splice(i,1);
            }
        }

        for (let i = this.SyncEffect.length - 1; i >= 0; i--) {
            let node = this.SyncEffect[i];
            try{
                let result = node.Update();
                if(!node.IsDone)
                {
                    node.Duration+=TimeKit.DeltaTime;
                }
                else
                {
                    this.SyncEffect.splice(i,1);
                }
            }
            catch{
                this.SyncEffect.splice(i,1);
            }
        }
    }
}

export enum EffectOrder
{
    BOT = 0,
    MID = 1,
    TOP = 3,
}

File: Unity/Assets/Scripts/ModTheSpire/Effect/EffectKit.cs
using System.Collections.Generic;
using UnityEngine;

namespace ModTheSpire.Effect
{
    public class EffectKit
    {
        private static EffectKit _inst;
        public static EffectKit Inst => _inst ??= new EffectKit();

        private EffectKit()
        {
        }

        private bool Waiting { get; set; }
        public List<AbstractEffect> List { get; } = new();
        public void Play(AbstractEffect effect)
        {
            this.List.Add(effect);
        }

        public void Update()
        {
            for (var i = this.List.Count - 1; i >= 0; i--)
            {
                var node = this.List[i];
                try
                {
                    node.Update();
                    if (!node.IsDone)
                    {
                        node.Duration += Time.deltaTime;
                    }
                    else
                    {
                        node.Dispose();
                        this.List.RemoveAt(i);
                    }
                }
                catch
                {
                    node.Dispose();
                    this.List.RemoveAt(i);
                }
            }
        }
    }

    public enum EffectOrder
    {
        BOT = 0,
        MID = 1,
        TOP = 3,
    }
}